client/src/interface/errors.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+#[cfg(feature = "p2p")]
+pub use p2p_errors::*;
+
+use crate::state::snapshot;
+
+use actix::MailboxError;
+use std::fmt::Debug;
+use thiserror::Error as DeriveError;
+
+#[derive(DeriveError, Debug)]
+pub enum ActorError {
+    #[error("Error sending message to Actor: `{0}`")]
+    Mailbox(#[from] MailboxError),
+    #[error("Target Actor has not been spawned or was killed.")]
+    TargetNotSpawned,
+}
+
+#[derive(DeriveError, Debug)]
+pub enum WriteVaultError {
+    #[error("Actor Error: `{0}`")]
+    Actors(#[from] ActorError),
+    #[error("Fatal Engine Error: `{0}`")]
+    Engine(String),
+}
+
+#[derive(DeriveError, Debug)]
+pub enum ReadSnapshotError {
+    #[error("Actor Error: `{0}`")]
+    Actors(#[from] ActorError),
+    #[error("Read Snapshot Error: `{0}`")]
+    Read(#[from] snapshot::ReadError),
+}
+#[derive(DeriveError, Debug)]
+pub enum WriteSnapshotError {
+    #[error("Actor Error: `{0}`")]
+    Actors(#[from] ActorError),
+    #[error("Write Snapshot Error: `{0}`")]
+    Write(#[from] snapshot::WriteError),
+}
+
+impl From<MailboxError> for WriteVaultError {
+    fn from(e: MailboxError) -> Self {
+        WriteVaultError::Actors(e.into())
+    }
+}
+
+impl From<MailboxError> for ReadSnapshotError {
+    fn from(e: MailboxError) -> Self {
+        ReadSnapshotError::Actors(e.into())
+    }
+}
+
+impl From<MailboxError> for WriteSnapshotError {
+    fn from(e: MailboxError) -> Self {
+        WriteSnapshotError::Actors(e.into())
+    }
+}
+
+#[cfg(feature = "p2p")]
+mod p2p_errors {
+
+    use super::*;
+    use p2p::{DialErr, ListenErr, ListenRelayErr, OutboundFailure};
+    use std::io;
+
+    #[derive(DeriveError, Debug)]
+    pub enum SpawnNetworkError {
+        #[error("Actor Mailbox Error: `{0}`")]
+        ActorMailbox(#[from] MailboxError),
+
+        #[error("Error: Network already running.")]
+        AlreadySpawned,
```

Comment:
I find these types of errors a bit hard to classify. I wouldn't really consider it a fatal error, since a imo a fatal error is an unexpected failure from which you can not recover.  The `AlreadySpawned` error is an expected error if you don't follow the invariant that you have mentioned above. Ideally, the code that is responsible for spawning the network (could be your code, or the code of the end-application) should take care that this error can never happen, so that it can be treated as `unreachable`. 
Apart from that, depending on the intention of the caller, the error may not even matter in some cases, e.g. when you use it as an "spawn a new p2p-actor if there isn't one already" (though one can argue that using it like that isn't very clean - but still, it's allowed).

---
Maybe that's even something to also add to your list in the discussion https://github.com/iotaledger/identity.rs/discussions/460 on how to classify errors: **"Expected vs Unexpected"** errors. It would also apply for e.g. `ActorError::TargetNotFound` or `OutboundFailure::NotPermitted`, which are also errors where, depending on the implementation, you known for sure they can never occur.

> This also feels like an error that could potentially be avoided by encoding the state of the system in the type system, but I don't know the code base enough to judge properly.

That would be desirable yes, but not sure how to do that. Do you have an example from e.g. the identity code base where something like this is done?

Reply:
Depending on whether you define your invariant one way or another, it's expected (local) or unexpected (fatal), so I think it's another way to look at the same consideration, which is still useful!

But it's fair to say that this is more of a case where the user needs to define whether this is an invariant, and unwrap or map it to unreachable. 

> Do you have an example from e.g. the identity code base where something like this is done?

The basic idea is to call `spawn` which returns a, say, `StrongholdP2p` type. On that type, you can not call `spawn` again, and so you don't have this error. This discourages accidentally calling spawn twice, but if it happens anyway, this then returns a lower-level error such as `AdressInUse`.

I've tried to do this with the identity actor, where creating an `Actor` means you also start to [spawn the listener](https://github.com/iotaledger/identity.rs/blob/feat/identity-actor/identity-actor/src/actor/actor.rs#L95) in the background. It guarantees any actor type you have is always listening, and you can't forget to start listening or accidentally try to spawn a listener twice. (You can still forget to listen on an address, though, which is a different topic).

The _type system guidance_ section in chapter 3 of _Rust for Rustaceans_ talks about another example for state machines with more states, which may also be interesting to look at.